### ISSUE TITLE
List/Grids that can't actually scroll should warn that repaint boundaries are being added, or just avoid adding them.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#103068]
The app at the bottom of this issue adds repaint boundaries by default for the grid view, even though it is not scrollable by itself and is shrink wrapped. This means that some of the entries will be chosen as raster cache entries, but in reality the whole thing could/should be raster cached.

The default behavior of adding repatint boundaries makes sense for grid views that actually scroll. It does not make sense when the grid view is just being used as a static layout mechanism (effectively a set of rows and columns). We should warn developers about this case and encourage them to turn off repaint boundaries for this case.

/cc @flar @jonahwilliams @goderbauer

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SizedBox(
          width: 350,
          child: ListView(
            children: <Widget>[
              Container(
                color: Colors.red,
                height: 45,
                width: 350,
                child: const Text('Header!'),
              ),
              const Divider(),
              Container(
                color: Colors.blue,
                height: 240,
                width: 350,
                child: const Text('Foo!'),
              ),
              const Divider(),
              GridView.count(
                crossAxisCount: 4,
                crossAxisSpacing: 20,
                mainAxisSpacing: 20,
                shrinkWrap: true,
                children: const <Widget>[
                  CircleAvatar(child: Text('AB')),
                  CircleAvatar(child: Text('CD')),
                  CircleAvatar(child: Text('EF')),
                  CircleAvatar(child: Text('GH')),
                  CircleAvatar(child: Text('BA')),
                  CircleAvatar(child: Text('DC')),
                  CircleAvatar(child: Text('FE')),
                  CircleAvatar(child: Text('HG')),
                ],
              ),
              const Divider(),
              Container(
                color: Colors.yellow,
                height: 140,
                width: 350,
                child: const Text('Bar!'),
              ),
              Container(
                color: Colors.green,
                height: 45,
                width: 350,
                child: const Text('Footer!'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

**Comment by dnfield** [comment]
This issue was discovered when looking at raster caching behavior in customer: money. That customer uses GridViews this way, and ends up with inefficient raster caching.

**Comment by jonahwilliams** [comment]
Here is a document discussing the performance in customer money: https://docs.google.com/document/d/1GlBVkJ99WY1X1wJ7xyCdsI_qhYaiQIRIITp_Snj3s9Q/edit?resourcekey=0-StI02EBd8ltGOcHlRvJYJg (sorry google only)

**Comment by goderbauer** [comment]
A better way to achieve what the example code is doing would be to have a CustomScrollView with a SliverList, a SliverGrid, and another SliverList following it. That also removes the need for the costly shrinkWrap. I haven't looked what this does to repaint boundaries, though.

**Comment by jonahwilliams** [comment]
FWIW in this application the UI thread is doing fine. The cost is entirely in what the framework is telling the engine to render

**Comment by dnfield** [comment]
The example code is mainly meant to demonstrate what I mean by a Grid view that's shrink wrapped and can't scroll. The rest can be ignored :)

**Comment by goderbauer** [comment]
The problem with the auto-added repaint boundaries would also stay the same, I just checked.

On a case-by-case basis, those repaint-boundaries can be disabled by setting `addRepaintBoundaries` to false on the GridView or SliverGridDelegate. That's not necessarily easy to discover, though.

**Comment by goderbauer** [comment]
> This means that some of the entries will be chosen as raster cache entries, but in reality the whole thing could/should be raster cached.

This is probably only true if the whole grid/list fits on one screen, though, right? If it doesn't, you'd be trashing the raster cache every time the outer ListView scrolls (at least in the sliver cases where the grid can actually lacily render its children in these situations).

**Comment by jonahwilliams** [comment]
You wouldn't thrash the cache in that case because it only applies after 3+ frames of stability, but you wouldn't get any benefits from it.

If you have a nested shrink wrapped list/grid, can it even scroll?

**Comment by dnfield** [comment]
> If you have a nested shrink wrapped list/grid, can it even scroll?

I think @goderbauer's point is that it could extend beyond the visible portion of the viewport it is in.

> On a case-by-case basis, those repaint-boundaries can be disabled by setting addRepaintBoundaries to false on the GridView or SliverGridDelegate. That's not necessarily easy to discover, though, and also requires manual work.

The request in this bug is that an assertion be added to the framework that would fire if a developer misses this. I believe it should be discoverable in the example provided, and if an assertion fired I would know to go back and disable repaint boundaries in my widget.

**Comment by jonahwilliams** [comment]
It doesn't mattter if it extends beyond the visible viewport though, only if it can be internally scrolled. If it can't, then it is already protected by the repaint boundary from the parent widget.

Assertion would be fine, but I also think we should change how the default is calculated

**Comment by goderbauer** [comment]
> It doesn't mattter if it extends beyond the visible viewport though, only if it can be internally scrolled. If it can't, then it is already protected by the repaint boundary from the parent widget.

What do you mean by internally scrolled? If you refactor this to actually use slivers, the SliverGrid will participate in regular scrolling and only paint what's necessary making it more efficient _if_ the grid is larger than the viewport. I am not sure if any of that applies in the awkward shrinkWrap solution presented above, though.

**Comment by goderbauer** [comment]
FWIW, I am not opposed to making this smarter. (And ideally it should just automatically do the right thing instead of asserting and requiring manual work). I just want to make sure that we cover the relevant cases.

**Comment by jonahwilliams** [comment]
I'll be in the office shortly, lets discuss in person?

**Comment by jonahwilliams** [comment]
So open questions:

Can we detect, from a SliverGrid/List that you are shrinkwrapped in another viewport? i.e. try and look up some scrollable specific inherited widget. If so, we could assert and ask that repaint boundaries are disabled - though there might be cases where that shouldn't be the default.

Instead we might want to make the repaint boundary behavior conditional. That would be difficult as we'd have to detect what the user has asked vs what the default is since there is no easy way to tell if an argument has been passed or not passed.

We could also make a reduce test case using dan's example above, though we might need to make some of the widgets more complex

**Comment by Piinks** [comment]
Joining in from @dnfield's comment here: https://discord.com/channels/608014603317936148/608021234516754444/971496850307809351

Most often I've seen folks use shrinkwrap for
- Putting a List or Grid view in a Column so it is not unbounded
- when nesting inside of another scroll view
  - in this case, the inner scrollview is the one that uses shrinkwrap and is not scrollable, using NeverScrollableScrollPhysics. This allows the content of that inner list to scrolls with the outer as if they are one list. The example above does not change the physics here, so the difference is that the gRidView can be scrolled independently from the ListView. See below.

<details>
<summary>Demo </summary>

https://user-images.githubusercontent.com/16964204/166819188-57b8c600-4848-4f11-990b-9954a2ddb7b3.mov

</details>

It's not ideal though, and @goderbauer's comment above about composing with slivers is much better for both cases, but developers seem to prefer just setting shrinkwrap rather than refactoring to use slivers.

We've been working on making them aware of what shrinkwrap acutally means and provide codelabs on composing with slivers, like here: https://www.youtube.com/watch?v=LUqDNnv_dh0&t=8s 

There is a dartpad workshop link in the comments.

I know customer: money uses a lot of shrink wrapping. I tried to refactor some of their code to see if there could be some performance improvements a few months ago, but wasn't able to prove it was costly for them at the time.

We may be able to conditionally apply repaint boundaries based on the dimensions of the viewport and whether or not it can scroll. One edge case I can think of though is on desktop and web where the user can resize the window. Something that did not scroll previously can find itself in a very small window and need to scroll.

**Comment by jonahwilliams** [comment]
If something has neverscrollablephysics, it definitely can't scroll right?

**Comment by jonahwilliams** [comment]
> I know customer: money uses a lot of shrink wrapping. I tried to refactor some of their code to see if there could be some performance improvements a few months ago, but wasn't able to prove it was costly for them at the time.

Yeah, this is one of those things that changes rasterization characteristics, but if you're not able to remove all of the extra layers (not just repaint boundary, but physical model and a few others), you will see 0 benefit.

**Comment by dnfield** [comment]
> I know customer: money uses a lot of shrink wrapping. I tried to refactor some of their code to see if there could be some performance improvements a few months ago, but wasn't able to prove it was costly for them at the time.

Some of that cost is coming right now not out of build times, but out of rasterization times. @jonahwilliams has more context there.

**Comment by dnfield** [comment]
I think the only time shrinkwrap makes sense are the cases where _actually_ having a nested scrollable makes sense, e.g. a text field within a longer column (like the text field I'm typing into right now on this web page).

I think if we asserted when you have a shrink-wrapped scrollable that can't actually scroll, we'd be in better shape.

**Comment by jonahwilliams** [comment]
In that case, it would likely be better to use a column/row/wrap as it wouldn't break up the picture

**Comment by Piinks** [comment]
> If something has neverscrollablephysics, it definitely can't scroll right?

Correct!

> I think the only time shrinkwrap makes sense are the cases where actually having a nested scrollable makes sense, e.g. a text field within a longer column (like the text field I'm typing into right now on this web page).

Not necessarily. Only when giving it unbounded height. In you TextField case, the scroll view would likely be contianed within a SizedBox or some sort. No need to shrinkwrap if it is given bounds.

**Comment by dnfield** [comment]
So what's a good motivating case for where you absolutely need to shrinkwrap?

**Comment by Piinks** [comment]
I don't know. I do not know what the original motivation for it was.
